### PR TITLE
Refactor rendering for testing

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- This refactors `core/render` to make it easier to test. The goal is to be able to create rendering tests directly in JavaScript without the need for fixtures.
- Main change is that `core/render` now has 2 new types:
  - `Environment` is an object that contains things that do not change between requests, such as the logging options, the routeCache, the markdown options, etc.
  - `RenderContext` is an object that contains things specific to a request, such as the request object and the scripts/styles that need to be included.

## Testing

- Some basic tests are added, more to come with bugs.
- Fixture tests should all pass as no external API changes occurred.

## Docs

N/A, internal refactor

### CODE REVIEW

File: packages/astro/test/units/render/jsx.test.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+				renderers: [await loadJSXRenderer()]
+			});
+		})
+
+		it('Can render slots', async () => {
+			const Wrapper = createComponent((result, _props, slots = {}) => {
+				return render`<div>${renderSlot(result, slots['myslot'])}</div>`;
+			});
+	
+			const Page = createAstroJSXComponent(() => {
```
Comment (matthewp) [inline packages/astro/test/units/render/jsx.test.js:27]: Can create an Astro JSX component in plain JS now!

File: packages/astro/test/units/render/jsx.test.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+		let env;
+		before(async () => {	
+			env = createBasicEnvironment({
+				logging,
+				renderers: [await loadJSXRenderer()]
+			});
+		})
+
+		it('Can render slots', async () => {
+			const Wrapper = createComponent((result, _props, slots = {}) => {
```
Comment (matthewp) [inline packages/astro/test/units/render/jsx.test.js:23]: Can create Astro components in plain JS now.

**Comment by matthewp** [comment]
!bench

**Comment by matthewp** [comment]
Nice! ~1k more requests in the same time. I think this comes down to the fact that we are now setting up this `Environment` object once when the server first starts and that gets reused by every request. That object has a `resolve` method which previously created a new closure on every request and no longer does.

**Review by natemoo-re — APPROVED** [review]
This is amazing, great work! Excited to build on this with our public testing API. ❤️

**Comment by matthewp** [comment]
I think we can do more to make this nice from a public API perspective. Probably most of the params we treat as required should not actually be required. We want them to be required internally so we get things like logging, but a public API would probably only need the request object and the page component. We can improve from here though

**Pull request merged**